File: go/0033.search-in-rotated-sorted-array/solution.go
// Created by Olgierd Palasz at 2026/05/06 21:33
// leetgo: dev
// https://leetcode.com/problems/search-in-rotated-sorted-array/

package main

import (
	"bufio"
	"fmt"
	"os"

	. "github.com/j178/leetgo/testutils/go"
)

// @lc code=begin

func search(nums []int, target int) int {
	left := 0
	right := len(nums) - 1

	for left <= right {
		mid := int((left + right) / 2)
		if target == nums[mid] {
			return mid
		}
		if nums[left] <= nums[mid] {
			if target > nums[mid] || target < nums[left] {
				left = mid + 1
			} else {
				right = mid - 1
			}
		} else {
			if target > nums[right] || target < nums[mid] {
				right = mid - 1
			} else {
				left = mid + 1
			}
		}
	}
	return -1
}

// @lc code=end

func main() {
	stdin := bufio.NewReader(os.Stdin)
	nums := Deserialize[[]int](ReadLine(stdin))
	target := Deserialize[int](ReadLine(stdin))
	ans := search(nums, target)

	fmt.Println("\noutput:", Serialize(ans))
}

File: rust/src/0033.search-in-rotated-sorted-array/solution.rs
// Created by Olgierd Palasz at 2026/08/10 22:34
// leetgo: dev
// https://leetcode.com/problems/search-in-rotated-sorted-array/

use anyhow::Result;
use leetgo_rs::*;

struct Solution;

// @lc code=begin

impl Solution {
    pub fn search(nums: Vec<i32>, target: i32) -> i32 {
        let mut left = 0;
        let mut right = nums.len() - 1;

        while left < right {
            let mid = left + (right - left) / 2;
            if nums[mid] > nums[right] {
                left = mid + 1
            } else {
                right = mid;
            }
        }

        let smallest = left;

        if target <= nums[nums.len() - 1] {
            left = smallest;
            right = nums.len();
        } else {
            left = 0;
            right = smallest;
        }

        while left < right {
            let mid = left + (right - left) / 2;
            if target < nums[mid] {
                right = mid;
            } else if target > nums[mid] {
                left = mid + 1;
            } else {
                return mid as i32;
            }
        }
        -1
    }
}

// @lc code=end

fn main() -> Result<()> {
    let nums: Vec<i32> = deserialize(&read_line()?)?;
    let target: i32 = deserialize(&read_line()?)?;
    let ans: i32 = Solution::search(nums, target).into();

    println!("\noutput: {}", serialize(ans)?);
    Ok(())
}
